### PR TITLE
ud am_zcopy implementation

### PR DESCRIPTION
**Pull request by alex-mikheev**
@yosefe  pls take a look

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1051/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1663/ for details.

File: src/uct/ib/ud/accel/ud_mlx5.c [hunk 1]
```diff
@@ -225,6 +225,75 @@ static ssize_t uct_ud_mlx5_ep_am_bcopy(uct_ep_h tl_ep, uint8_t id,
...
+    struct ibv_mr *mr = memh;
+    uint32_t lkey;
+    struct mlx5_wqe_ctrl_seg *ctrl;
+    struct mlx5_wqe_inl_data_seg *inl;
+    struct mlx5_wqe_data_seg *dptr;
+    unsigned wqe_size;
+    uct_ud_neth_t *neth;
+
+
+    UCT_CHECK_LENGTH(sizeof(uct_ud_neth_t) + header_length,
```
Comment (yosefe) [inline src/uct/ib/ud/accel/ud_mlx5.c]: need to check payload size too

File: src/uct/ib/ud/accel/ud_mlx5.c [hunk 2]
```diff
@@ -225,6 +225,75 @@ static ssize_t uct_ud_mlx5_ep_am_bcopy(uct_ep_h tl_ep, uint8_t id,
...
+        }
+        wqe_size += sizeof(*dptr);
+        uct_ib_mlx5_set_data_seg(dptr, payload, length, lkey);
+    }
+
+    UCT_UD_EP_HOOK_CALL_TX(&ep->super, neth);
+    uct_ud_mlx5_post_send(iface, ep, ctrl, wqe_size);
+
+    skb->len = sizeof(*neth) + header_length;
+    memcpy(skb->neth, neth, sizeof(*neth));
```
Comment (yosefe) [inline src/uct/ib/ud/accel/ud_mlx5.c]: can this be done with single memcpy?

File: src/uct/ib/ud/base/ud_def.h [hunk 3]
```diff
@@ -135,6 +136,27 @@ typedef struct uct_ud_send_skb {
     uct_ud_neth_t           neth[0];
 } UCS_S_PACKED uct_ud_send_skb_t;
 
+typedef struct uct_ud_zcopy_hdr {
```
Comment (yosefe) [inline src/uct/ib/ud/base/ud_def.h]: maybe not called it "header"? because it's not a packet header

File: src/uct/ib/ud/base/ud_ep.h [hunk 4]
```diff
@@ -198,6 +198,8 @@ struct uct_ud_ep {
          ucs_queue_head_t       window;       /* send window: [acked_psn+1, psn-1] */
          uct_ud_ep_pending_op_t pending;      /* pending ops */
          ucs_time_t             send_time;    /* tx time of last packet */
+         int                    zcopy_comps;  /* true if there are zcopy completions that
```
Comment (yosefe) [inline src/uct/ib/ud/base/ud_ep.h]: can we add it to endpoint flags?

File: src/uct/ib/ud/base/ud_iface.c [hunk 5]
```diff
@@ -373,7 +374,7 @@ UCS_CLASS_INIT_FUNC(uct_ud_iface_t, uct_iface_ops_t *ops, uct_pd_h pd,
     }
 
     status = uct_iface_mpool_init(&self->super.super, &self->tx.mp,
-                                sizeof(uct_ud_send_skb_t) + 4096 /* TODO mtu */,
+                                sizeof(uct_ud_zcopy_hdr_t) + sizeof(uct_ud_send_skb_t) + 4096 /* TODO mtu */,
```
Comment (yosefe) [inline src/uct/ib/ud/base/ud_iface.c]: IMHO it should be ucs_max(sizeof(zcopy_hdr), 4096)

File: src/uct/ib/ud/accel/ud_mlx5.c (hunk 2 above)
Comment (shamisp) [inline src/uct/ib/ud/accel/ud_mlx5.c]: unrelated, but what do u think about introduction of memcpy within UCX ?

Comment (yosefe) [inline src/uct/ib/ud/accel/ud_mlx5.c]: assuming it would be more optimized than gcc's?

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1073/ for details (Mellanox internal link).

**Comment by yosefe** [comment]
looks good, will merge after tests pass

File: src/uct/ib/ud/accel/ud_mlx5.c (hunk 2 above)
Comment (shamisp) [inline src/uct/ib/ud/accel/ud_mlx5.c]: @yosefe - correct. It seems that for arm the glibc version is not the best, for 64bit platforms

Comment (yosefe) [inline src/uct/ib/ud/accel/ud_mlx5.c]: ok, so i guess we should add ucs_memcpy to ucs/arch and define it to memcpy in x86_64 and the optimized version on arm. it should probably be an inline function.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1682/ for details.

**Comment by yosefe** [comment]
bot:retest

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1077/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1685/ for details.

**Comment by yosefe** [comment]
:+1:

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1687/ for details.

**Pull request merged**